gear-programs/historical-proxy/app/src/service.rs

```diff
@@ -0,0 +1,159 @@
+    Relayed {
+        slot: u64,
+        block_number: u64,
+        transaction_index: u32,
+    },
+}
+
+pub struct HistoricalProxyService<'a, ExecContext> {
+    state: &'a RefCell<ProxyState>,
+    exec_context: ExecContext,
+}
+
+#[sails_rs::service(events = Event)]
+impl<'a, T> HistoricalProxyService<'a, T>
+where
+    T: ExecContext,
+{
+    pub fn new(state: &'a RefCell<ProxyState>, exec_context: T) -> Self {
+        Self {
+            state,
+            exec_context,
+        }
+    }
+
+    pub fn config(&self) -> Config {
+        self.state.borrow().config
+    }
+
+    pub fn update_config(&mut self, config_new: Config) -> Result<(), ProxyError> {
+        let source = self.exec_context.actor_id();
+        let mut state = self.state.borrow_mut();
+        if source != state.admin {
+            return Err(ProxyError::NotAdmin);
+        }
+
+        state.config = config_new;
+        Ok(())
+    }
+
+    pub fn admin(&self) -> ActorId {
+        self.state.borrow().admin
+    }
+
+    pub fn endpoint_for(&mut self, slot: Slot) -> Result<ActorId, ProxyError> {
+        self.state.borrow().endpoints.endpoint_for(slot)
+    }
+
+    pub fn add_endpoint(&mut self, slot: Slot, endpoint: ActorId) -> Result<(), ProxyError> {
+        let source = self.exec_context.actor_id();
+
+        let mut state = self.state.borrow_mut();
+        if source != state.admin {
+            return Err(ProxyError::NotAdmin);
+        }
+
+        state.endpoints.push(slot, endpoint);
+        Ok(())
+    }
+
+    pub fn endpoints(&self) -> Vec<(Slot, ActorId)> {
+        self.state.borrow().endpoints.endpoints()
+    }
+
+    /// Redirect message to ERC20 Relay service which is valid for `slot`.
+    /// If message is relayed successfully then reply from relay service is sent to
+    /// `vft_gateway` address and proofs are returned.
+    ///
+    /// # Parameters
+    ///
+    /// - `slot`: slot for which message is relayed.
+    /// - `tx_index`: transaction index for message.
+    /// - `proofs`: SCALE encoded `EthToVaraEvent`.
+    /// - `client`: client address to send receipt to on success.
+    /// - `client_route`: route to send receipt to on success.
+    /// # Returns
+    /// - `(Vec<u8>, Vec<u8>)`: on success where first vector is receipt and second vector is reply from calling `client_route`.
+    /// - `ProxyError`: if redirect failed
+    ///
+    #[allow(clippy::await_holding_refcell_ref)]
+    pub async fn redirect(
+        &mut self,
+        slot: Slot,
+        tx_index: u64,
+        proofs: Vec<u8>,
+        client: ActorId,
+        client_route: Vec<u8>,
+    ) -> Result<(Vec<u8>, Vec<u8>), ProxyError> {
+        let state = self.state.borrow();
+        let reply_deposit = state.config.reply_deposit;
+        let reply_timeout = state.config.reply_timeout;
+        let endpoint = state.endpoints.endpoint_for(slot)?;
+        drop(state);
+        let event = EthToVaraEvent::decode(&mut proofs.as_slice()).map_err(|e| {
```

Comment:
No, it shouldn't be done here. Instead send the payload the same way as `submit_receipt` using the `erc20_relay` route from IDL